src/libsystemd/sd-device/device-enumerator.c

```diff
@@ -886,25 +888,33 @@ static int parent_crawl_children(sd_device_enumerator *enumerator, const char *p
                 if (k < 0)
                         r = k;
 
-                if (maxdepth > 0)
-                        parent_crawl_children(enumerator, child, maxdepth - 1);
-                else
-                        log_debug("sd-device-enumerator: Max depth reached, %s: ignoring devices", child);
+                k = set_ensure_consume(stack, &path_hash_ops_free, TAKE_PTR(child));
+                if (k < 0)
+                        r = k;
         }
 
         return r;
 }
 
 static int enumerator_scan_devices_children(sd_device_enumerator *enumerator) {
+        _cleanup_set_free_ Set *stack = NULL;
         const char *path;
         int r = 0, k;
 
+        assert(enumerator);
+
         SET_FOREACH(path, enumerator->match_parent) {
                 k = parent_add_child(enumerator, path);
                 if (k < 0)
                         r = k;
 
-                k = parent_crawl_children(enumerator, path, DEVICE_ENUMERATE_MAX_DEPTH);
+                k = parent_crawl_children(enumerator, path, &stack);
+                if (k < 0)
+                        r = k;
+        }
+
+        for (char *p; (p = set_steal_first(stack)); free(p)) {
```

Comment:
doesn't matter, but I must admit I am not too much of a fan of using for loops in such an adventurous way. It's fine if hidden in a macro, but in first class code, meh. If I look at this line I need to think three times before I grok that this isn't a memory leak here.


If you'd write it like this:

```c
for (;;) {
        _cleanup_free char *p = set_steal_first(stack);
        if (!p)
                break;
        …
```

Then yes, it's four lines instead of one, but it has this major benefit that it's robust to leaving the loop via break/return early won't cause a memory leak.

anyway, this doesn't matter much, the code is not too complex yet, but this is a pitfall, and I like simple and obvious code...

Reply:
Thanks. That looks better for me. Addressed at #24609.